libdrgn/symbol.c

```diff
@@ -174,3 +178,291 @@ void drgn_symbol_result_builder_array(struct drgn_symbol_result_builder *builder
 	symbolp_vector_shrink_to_fit(&builder->vector);
 	symbolp_vector_steal(&builder->vector, syms_ret, count_ret);
 }
+
+static int name_compar(const void *lhs, const void *rhs, void *arg)
+{
+	struct drgn_symbol_index *ix = arg;
+	uint32_t left_ix = *(const uint32_t *)lhs;
+	uint32_t right_ix = *(const uint32_t *)rhs;
+	return strcmp(ix->symbols[left_ix].name, ix->symbols[right_ix].name);
+}
+
+static int addr_compar(const void *lhs, const void *rhs, void *args)
+{
+	const struct drgn_symbol *left = lhs;
+	const struct drgn_symbol *right = rhs;
+	// returning a simple subtraction would not work well since these are
+	// unsigned
+	if (left->address < right->address)
+		return -1;
+	else if (left->address > right->address)
+		return 1;
+	else
+		return 0;
+}
+
+struct drgn_error *
+drgn_symbol_index_init(struct drgn_symbol *symbols, uint32_t count,
+		       char *buffer, struct drgn_symbol_index *ret)
+{
+
+	ret->symbols = symbols;
+	ret->num_syms = count;
+	ret->strings = buffer;
+	ret->name_sort = malloc_array(count, sizeof(ret->name_sort[0]));
+	if (!ret->name_sort)
+		return &drgn_enomem;
+	ret->max_addrs = malloc_array(count, sizeof(ret->max_addrs[0]));
+	if (!ret->max_addrs) {
+		free(ret->name_sort);
+		return &drgn_enomem;
+	}
+	drgn_symbol_name_table_init(&ret->htab);
+
+	// In many cases (e.g kallsyms), symbols are already sorted by address,
+	// but not always.  Check whether sorted, and if not, sort.
+	for (uint32_t i = 1; i < ret->num_syms; i++) {
+		if (ret->symbols[i - 1].address > ret->symbols[i].address) {
+			qsort_r(ret->symbols, count, sizeof(ret->symbols[0]), addr_compar, NULL);
+			break;
+		}
+	}
+
+	// Kallsyms doesn't include symbol lengths, so symbols are
+	// non-overlapping. But this is not true in general! Symbols may
+	// overlap, which makes address lookup complicated. Rather than using a
+	// complex range data structure, we can use two binary searches, one to
+	// find the first symbol which could overlap with an address, and one to
+	// find the last symbol, and then linearly search that array. This
+	// performs poorly if there are symbols which span many others, but
+	// that's a rare case. In order to do this strategy, we need an array
+	// that contains the maximum address spanned by any symbol at or before
+	// that index.
+	ret->max_addrs[0] = ret->symbols[0].address + ret->symbols[0].size;
+	for (uint32_t i = 1; i < ret->num_syms; i++) {
+		uint64_t max_addr = ret->symbols[i].address + ret->symbols[i].size;
+		ret->max_addrs[i] = max(ret->max_addrs[i - 1], max_addr);
+	}
+
+	// Sort the "name_sort" array by name so we get runs of symbols with the
+	// same name
+	for (uint32_t i = 0; i < ret->num_syms; i++)
+		ret->name_sort[i] = i;
+	qsort_r(ret->name_sort, ret->num_syms, sizeof(ret->name_sort[0]),
```

Comment:
I liked the helper function you wrote on that issue a lot since it makes the code a bit easier to read. So I added it under the name `qsort_arg` and converted existing users. But if you'd prefer I just use a thread local directly without an added helper function, I can switch it :)